editor/plugins/gradient_editor_plugin.cpp

```diff
@@ -31,27 +31,583 @@
 #include "gradient_editor_plugin.h"
 
 #include "canvas_item_editor_plugin.h"
+#include "core/os/keyboard.h"
 #include "editor/editor_node.h"
 #include "editor/editor_scale.h"
 #include "editor/editor_settings.h"
+#include "editor/editor_spin_slider.h"
 #include "editor/editor_undo_redo_manager.h"
 #include "node_3d_editor_plugin.h"
+#include "scene/gui/flow_container.h"
+#include "scene/gui/separator.h"
 
-void GradientReverseButton::_notification(int p_what) {
+int GradientEditorRect::get_point_at(int x) {
+	int result = -1;
+	int total_w = get_gradient_rect_width();
+	float min_distance = 1e20;
+	for (int i = 0; i < gradient->get_point_count(); i++) {
+		// Ignore points outside of [0, 1].
+		if (gradient->get_offset(i) < 0) {
+			continue;
+		} else if (gradient->get_offset(i) > 1) {
+			break;
+		}
+		// Check if we clicked at point.
+		float distance = ABS(x - gradient->get_offset(i) * total_w);
+		float min = handle_width * 0.8; // Allow the mouse to be more than half a handle width away for ease of grabbing.
+		if (distance <= min && distance < min_distance) {
+			result = i;
+			min_distance = distance;
+		}
+	}
+	return result;
+}
+
+int GradientEditorRect::get_index_at_offset(float p_offset) {
+	int result = 0;
+	while (result < gradient->get_point_count() && gradient->get_offset(result) < p_offset) {
+		result++;
+	}
+	return result;
+}
+
+int GradientEditorRect::get_gradient_rect_width() {
+	return get_size().width - get_size().height - draw_spacing - handle_width;
+}
+
+void GradientEditorRect::show_color_picker() {
+	ERR_FAIL_INDEX(selected_index, gradient->get_point_count());
+
+	picker->set_pick_color(gradient->get_color(selected_index));
+	Size2 minsize = popup->get_contents_minimum_size();
+	float viewport_height = get_viewport_rect().size.y;
+
+	// Determine in which direction to show the popup. By default popup below.
+	// But if the popup doesn't fit below and the Gradient Editor is in the bottom half of the viewport, show above.
+	bool show_above = false;
+	if (get_global_position().y + get_size().y + minsize.y > viewport_height && get_global_position().y * 2 + get_size().y > viewport_height) {
+		show_above = true;
+	}
```

Comment:
Can be simplified to
```suggestion
	bool show_above = (get_global_position().y + get_size().y + minsize.y > viewport_height && get_global_position().y * 2 + get_size().y > viewport_height);
```